sql/expression/collatedexpr.go

```diff
@@ -58,14 +58,25 @@ func (ce *CollatedExpression) IsNullable() bool {
 
 // Type implements the sql.Expression interface.
 func (ce *CollatedExpression) Type() sql.Type {
-	return ce.expr.Type()
+	typ := ce.expr.Type()
+	if collatedType, ok := typ.(sql.TypeWithCollation); ok {
+		return collatedType.WithNewCollation(ce.collation)
+	}
+	// If this isn't a collated type then this should fail, as we can't apply a collation to an expression that does not
+	// have a charset. However, since Type() doesn't fail, we'll instead check in Eval().
+	return typ
```

Comment:
we can't fail here, but we could catch this case in the constructor

Reply:
We can't quite check this in the constructor either. Expressions such as unresolved columns won't be resolved yet during the parsing phase, so we can't guarantee much until we're evaluating the expression. I'll update the comment to reflect this.